storage/gcp/gcp.go

```diff
@@ -154,58 +245,53 @@ func (s *Storage) initDB(ctx context.Context) error {
 	return nil
 }
 
-// tileSuffix returns either the empty string or a tiles API suffix based on the passed-in tile size.
-func tileSuffix(s uint64) string {
-	if p := s % 256; p != 0 {
-		return fmt.Sprintf(".p/%d", p)
-	}
-	return ""
-}
-
-// setTile stores a tile in GCS.
+// assignEntries durably assigns each of the passed-in entries an index in the log.
 //
-// The location to which the tile is written is defined by the tile layout spec.
-func (s *Storage) setTile(ctx context.Context, level, index uint64, tile [][]byte) error {
-	tileSize := uint64(len(tile))
-	klog.V(2).Infof("StoreTile: level %d index %x ts: %x", level, index, tileSize)
-	if tileSize == 0 || tileSize > 256 {
-		return fmt.Errorf("tileSize %d must be > 0 and <= 256", tileSize)
+// Entries are allocated contiguous indices, in the order in which they appear in the entries parameter.
+// This is achieved by storing the passed-in entries in the Seq table in Spanner, keyed by the
+// index assigned to the first entry in the batch.
+func (s *spannerSequencer) assignEntries(ctx context.Context, entries [][]byte) (uint64, error) {
+	// Flatted the entries into a single slice of bytes which we can store in the Seq.v columne
+	b := &bytes.Buffer{}
+	e := gob.NewEncoder(b)
+	if err := e.Encode(entries); err != nil {
+		return 0, fmt.Errorf("failed to serialise batch: %v", err)
 	}
-	t := slices.Concat(tile...)
+	data := b.Bytes()
+	num := len(entries)
 
-	// Pass an empty rootDir since we don't need this concept in GCS.
-	tPath := layout.TilePath(level, index) + tileSuffix(tileSize)
+	var next int64 // Unfortunately, Spanner doesn't support uint64 so we'll have to cast around a bit.
 
-	return s.objStore.setObject(ctx, tPath, t, &gcs.Conditions{DoesNotExist: true})
-}
+	_, err := s.dbPool.ReadWriteTransaction(ctx, func(ctx context.Context, txn *spanner.ReadWriteTransaction) error {
+		// First we need to grab the next available sequence number from the SeqCoord table.
+		row, err := txn.ReadRowWithOptions(ctx, "SeqCoord", spanner.Key{0}, []string{"id", "next"}, &spanner.ReadOptions{LockHint: spannerpb.ReadRequest_LOCK_HINT_EXCLUSIVE})
+		if err != nil {
+			return err
+		}
+		var id int64
+		if err := row.Columns(&id, &next); err != nil {
+			return err
+		}
 
-// getTile returns the tile at the given tile-level and tile-index for the specified log size, or
-// an error if it doesn't exist.
-func (s *Storage) getTile(ctx context.Context, level, index, logSize uint64) ([][]byte, error) {
-	tileSize := layout.PartialTileSize(level, index, logSize)
+		next := uint64(next) // Shadow next with a uint64 version of the same value to save on casts.
+		m := []*spanner.Mutation{
+			// Insert our newly sequenced batch of entries into Seq,
+			spanner.Insert("Seq", []string{"id", "seq", "v"}, []interface{}{0, int64(next), data}),
```

Comment:
Would it be worth splitting data bundles, and aligning their end on a multiple of 256?
I expect that integration will work in batches as well "integrate X bundles at most" rather than "integrate all available bundles in one go"? In which case, splitting data bundles will allow to align integration with full leaf bundles and limit the number of partial leaf bundles / L0 tiles.

Reply:
Alternatively - did you try sequencing entries into bundles directly?

Reply:
Yeah, that's (aligning to tile size) worth exploring, I'll pop a TODO to come back and try that out later when we have something working.

Sequencing to bundles is something I tried very early on, the main problem you run into with that approach (on a non-transactional storage), is that given a tree of size N, it's very likely that two frontends will want to add a different number of leaves to the sequenced bundles at the same time. At which point you can't rely on GCS preconditions to save you, since the objects have different keys (because partial suffix), so then you need to either a) somehow create a global lock, or b) rely on some transactional infra to manage the sequencing.

Each has trade-offs, for now a transactional "coordination" is the right way to go (for scale), although I am keeping a back-of-mind thread thinking about how to also scale down for "Tessera serverless" use cases.